src/lib.rs

```diff
@@ -84,30 +72,539 @@ fn serialize_from_raw_pointer<T: ToBytes>(
+    let heap_obj = Box::new(cmt);
+    println!("Ctor: heap_obj {:p}", heap_obj);
+    let raw_obj = Box::into_raw(heap_obj);
+    println!("      raw_obj  {:p}", raw_obj);
+    raw_obj
+}
+
+#[no_mangle]
+pub extern "C" fn zendoo_commitment_tree_delete(ptr : *mut CommitmentTree) {
+    if ptr.is_null() {
+        println!("delete: nullptr!");
+        return;
+    }
+    unsafe {
+        println!("Dtor: raw_obj  {:p}", ptr);
+        let heap_obj = Box::from_raw(ptr);
+        println!("      heap_obj {:p}", heap_obj);
+    }
+}
+
+
+#[no_mangle]
+pub extern "C" fn zendoo_commitment_tree_add_scc(ptr : *mut CommitmentTree,
+    sc_id: *mut BufferWithSize,
+    amount: i64,
+    pub_key: *mut BufferWithSize,
+    withdrawal_epoch_length: u32,
+    custom_data: *mut BufferWithSize,
+    constant:    *mut BufferWithSize,
+    cert_vk: *mut BufferWithSize,
+    btr_vk: *mut BufferWithSize,
+    csw_vk: *mut BufferWithSize,
+    tx_hash: *mut BufferWithSize,
+    out_idx: u32, 
+    ret_code : &mut CctpErrorCode)-> bool
+{
+    *ret_code = CctpErrorCode::OK;
+    if ptr.is_null() {
+        *ret_code = CctpErrorCode::NullPtr;
+        dbg!(*ret_code);
+        return false;
+    }
+
+    // optional or variable size parameters
+    let mut rs_custom_data : &[u8] = &[];
+    let (is_ok, _err) = check_buffer(custom_data);
+    if is_ok {
+        rs_custom_data = unsafe {
+            if (*custom_data).len > CUSTOM_DATA_MAX_SIZE { *ret_code = CctpErrorCode::InvalidBufferLength; dbg!(*ret_code); return false; } 
+            slice::from_raw_parts((*custom_data).data, (*custom_data).len)
+        };
+    }
+
+    let mut rs_constant : Option<&[u8]> = None;
+    let (is_ok, _err) = check_buffer(constant);
+    if is_ok {
+        rs_constant = Some(unsafe {
+            if (*constant).len != FIELD_SIZE { *ret_code = CctpErrorCode::InvalidBufferLength; dbg!(*ret_code); return false; } 
+            slice::from_raw_parts((*constant).data, (*constant).len)
+        });
+    }
+
+    let mut rs_csw_vk : Option<&[u8]> = None;
+    let (is_ok, _err) = check_buffer(csw_vk);
+    if is_ok {
+        rs_csw_vk = Some(unsafe {
+            if (*csw_vk).len != SC_VK_SIZE { *ret_code = CctpErrorCode::InvalidBufferLength; dbg!(*ret_code); return false; } 
+            slice::from_raw_parts((*csw_vk).data, (*csw_vk).len)
+        });
+    }
+
+    let mut rs_btr_vk : Option<&[u8]> = None;
+    let (is_ok, _err) = check_buffer(btr_vk);
+    if is_ok {
+        rs_btr_vk = Some(unsafe {
+            if (*btr_vk).len != SC_VK_SIZE { *ret_code = CctpErrorCode::InvalidBufferLength; dbg!(*ret_code); return false; } 
+            slice::from_raw_parts((*btr_vk).data, (*btr_vk).len)
+        });
+    }
+
+    // mandatory and constant size parameters
+    let (is_ok, err) = check_buffer_length(sc_id, UINT_256_SIZE);
+    if !is_ok { *ret_code = err; dbg!(err); return false; }
+    let rs_sc_id = &mut (unsafe { slice::from_raw_parts((*sc_id).data,   (*sc_id).len)});
```

Comment:
Here as well as in further occurrences there is no need for casting to `&mut` due to `cmt.add_scc` method takes `rs_sc_id` as immutable.
Also for a mutable slice there is `from_raw_parts_mut` method.

Reply:
Ok, done